applications/trilinos_application/external_includes/mpi_amgcl_ns_solver.h

```diff
@@ -0,0 +1,391 @@
+    * when solving a mixed u-p problem, it is important to identify the row associated to v and p.
+    * another example is the automatic prescription of rotation null-space for smoothed-aggregation solvers
+    * which require knowledge on the spatial position of the nodes associated to a given dof.
+    * This function tells if the solver requires such data
+    */
+    virtual bool AdditionalPhysicalDataIsNeeded()
+    {
+        return true;
+    }
+
+    /** Some solvers may require a minimum degree of knowledge of the structure of the matrix. To make an example
+     * when solving a mixed u-p problem, it is important to identify the row associated to v and p.
+     * another example is the automatic prescription of rotation null-space for smoothed-aggregation solvers
+     * which require knowledge on the spatial position of the nodes associated to a given dof.
+     * This function is the place to eventually provide such data
+     */
+    void ProvideAdditionalData (
+        SparseMatrixType& rA,
+        VectorType& rX,
+        VectorType& rB,
+        typename ModelPart::DofsArrayType& rdof_set,
+        ModelPart& r_model_part
+    )
+    {
+KRATOS_WATCH(__LINE__)
+        int my_pid = rA.Comm().MyPID();
+
+        //filling the pressure mask
+        if(mp.size() != static_cast<unsigned int>(rA.NumMyRows())) mp.resize( rA.NumMyRows() );
+KRATOS_WATCH(__LINE__)
+        int counter = 0;
+        for (ModelPart::DofsArrayType::iterator it = rdof_set.begin(); it!=rdof_set.end(); it++)
+        {
+            if( it->GetSolutionStepValue ( PARTITION_INDEX ) == my_pid )
+            {
+                mp[counter]  = (it->GetVariable().Key() == PRESSURE);
+                counter++;
+            }
+        }
+KRATOS_WATCH(__LINE__)
+        if(counter != rA.NumMyRows())
+            KRATOS_ERROR << "pressure mask as a size " << mp.size() << " which does not correspond with the number of local rows:" << rA.NumMyRows() << std::endl; 
+
+    }
+
+    /**
+     * Print information about this object.
+     */
+    void  PrintInfo ( std::ostream& rOStream ) const
+    {
+        rOStream << "AMGCL_MPI solver finished.";
+    }
+
+    /**
+     * Print object's data.
+     */
+    void  PrintData ( std::ostream& rOStream ) const
+    {
+    }
+
+private:
+
+    double mtol;
+    int mmax_it;
+    int mverbosity;
+    unsigned int mcoarse_enough;
+    bool muse_linear_deflation = false;
+    bool mprovide_coordinates = false;
+    bool muse_block_matrices_if_possible = false;
+
+    std::vector< char > mp; //pressure mask
+    boost::property_tree::ptree mprm;
+
+
+    /**
+     * Assignment operator.
+     */
+    AmgclMPI_NS_Solver& operator= ( const AmgclMPI_NS_Solver& Other );
+
+    /**
+     * Copy constructor.
+     */
+    AmgclMPI_NS_Solver ( const AmgclMPI_NS_Solver& Other );
+
+};
+
+
+/**
+ * input stream function
+ */
+template<class TSparseSpaceType, class TDenseSpaceType,class TReordererType>
```

Comment:
actually i also defined as deleted the copy constructor and the assignment operator